curvefs/src/client/xattr_manager.cpp

```diff
@@ -0,0 +1,539 @@
+        std::set<uint64_t> inodeIds;
+        std::list<InodeAttr> attrs;
+        auto tret = ConcurrentListDentry(&dentryList, iStack, stackMutex,
+                                         false, inflightNum, ret);
+        if (!tret) {
+            return;
+        }
+        {
+            std::lock_guard<std::mutex> guard(*stackMutex);
+            for (const auto &it : dentryList) {
+                if (it.type() == FsFileType::TYPE_DIRECTORY) {
+                    iStack->emplace(it.inodeid());
+                }
+                inodeIds.emplace(it.inodeid());
+            }
+            inflightNum->fetch_sub(1);
+        }
+        if (!inodeIds.empty()) {
+            auto tret = inodeManager_->BatchGetInodeAttr(&inodeIds, &attrs);
+            if (tret == CURVEFS_ERROR::OK) {
+                std::lock_guard<std::mutex> guard(*valueMutex);
+                for (const auto &it : attrs) {
+                    if (it.type() == FsFileType::TYPE_DIRECTORY) {
+                        summaryInfo->subdirs++;
+                    } else {
+                        summaryInfo->files++;
+                    }
+                    summaryInfo->entries++;
+                    summaryInfo->fbytes += it.length();
+                    // record hardlink
+                    if (it.type() != FsFileType::TYPE_DIRECTORY &&
+                        it.nlink() > 1) {
+                        std::lock_guard<std::mutex> guard(*mapMutex);
+                        auto iter = hardLinkMap->find(it.inodeid());
+                        if (iter != hardLinkMap->end()) {
+                            iter->second += it.length();
+                        } else {
+                            hardLinkMap->emplace(it.inodeid(), 0);
+                        }
+                    }
+                }
+            } else {
+                ret->store(false);
+                return;
+            }
+        }
+    }
+}
+
+CURVEFS_ERROR XattrManager::CalAllLayerSumInfo(Inode *inode) {
+    std::stack<uint64_t> iStack;
+    std::mutex stackMutex;
+
+    // record hard link, <inodeId, need2minus>
+    std::unordered_map<uint64_t, uint64_t> hardLinkMap;
+    std::mutex mapMutex;
+
+    SummaryInfo summaryInfo;
+    std::mutex valueMutex;
+
+    auto ino = inode->inodeid();
+    iStack.emplace(ino);
+    std::vector<Thread> threadpool;
+    Atomic<uint32_t> inflightNum(0);
+    Atomic<bool> ret(true);
+
+    for (auto i = listDentryThreads_; i > 0; i--) {
+        try {
+            threadpool.emplace_back(Thread(
+                &XattrManager::ConcurrentGetInodeAttr,
+                this, &iStack, &stackMutex, &hardLinkMap, &mapMutex,
+                &summaryInfo, &valueMutex, &inflightNum, &ret));
+        } catch (const std::exception& e) {
+            LOG(WARNING) << "CalAllLayerSumInfo create thread failed,"
+                         << " err: " << e.what();
+        }
+    }
+
+    if (threadpool.empty()) {
+        return CURVEFS_ERROR::INTERNAL;
```

Comment:
what happens when we return an error, does it will affect all io requests?

Reply:
The error returned here is same as getXattrInfo failed, the error will return as EIO, and not affect other io requests.